Make BbFlip on CPU act as on GPU

awolant opened the pull request:
CPU version of `BbFlip` was filtering boxes outside of the image. That's why it did not work with `SSDRandomCrop`.  Now both versions act the same in this manner. 

Signed-off-by: Albert Wolant <awolant@nvidia.com>

JanuszL commented:
Maybe dali/test/python/test_random_bbcrop.py could be extended (and possibly renamed) to cover this change as well.

awolant commented:
This change doesn't touch random cropping and there is no flip in dali/test/python/test_random_bbcrop.py 
I was planning to look exactly why ssd random crop gives boxes outside of images later and then maybe adjust this test, if necessary?
I can also maybe add python test for flipping?

awolant commented:
Build 675871

JanuszL commented:
> This change doesn't touch random cropping and there is no flip in dali/test/python/test_random_bbcrop.py
> I was planning to look exactly why ssd random crop gives boxes outside of images later and then maybe adjust this test, if necessary?
> I can also maybe add python test for flipping?

Sure. I just just thinking about testing CPU and GPU version side by side. Not necessarily using the same script as for bbox crop.

awolant commented:
Right. I will add this kind of test.

szalpal commented:
I'm wondering, in what circumstances we would like the BbFlip to work on bbs that are (even partially) outside of the image? 
Maybe we're should consider adding DALI_ENFORCE to runimpl guarding from this condition? 

I certainly wouldn't expect invalid bb on the input...

awolant commented:
I added different flips to `dali/test/python/test_random_bbcrop.py` and renamed it as `dali/test/python/test_detection_pipeline.py`
I will refactor this test in the separate PR - I will add other operators from detection pipeline (ex. box encoder) and add testing against reference implementations from PyTorch. This will give us pretty comprehensive test for SSD related operators.

I also fixed BbFlip on GPU for empty boxes - it used to throw `Floating point exception` due to division by 0.

jantonguirao commented on dali/test/python/test_detection_pipeline.py:
```diff
@@ -56,7 +72,7 @@ def define_graph(self):
         return self.base_define_graph(images, labels, bb)
 
-test_data = {
+test_data = {\
```
is that `\` intentional?

awolant replied:
No, removed it.

awolant commented:
Rebuild #681097

mzient commented on dali/pipeline/operators/geometric/bb_flip.cc:
```diff
@@ -75,7 +75,7 @@ void BbFlip<CPUBackend>::RunImpl(dali::SampleWorkspace *ws, const int idx) {
   for (int i = 0; i < input.size(); i += 4) {
-    auto bbox = ltrb_ ? BoundingBox::FromLtrb(&input_data[i])
+    auto bbox = ltrb_ ? BoundingBox::FromLtrb(&input_data[i], BoundingBox::NoBounds())
                       : BoundingBox::FromXywh(&input_data[i]);
```
Why the double standard here? It's more likely that an XYWH box will be out of bounds than LTRB due to rounding problems.

awolant replied:
Done

awolant commented:
Rebuild #682159
